Support latest MongoDB driver

williamlardier opened the pull request:
There is actually no change to make, but I seized the opportunity to cleanup a bit related logic, and the lib we use for testing.

The main contribution is to use MongoDB stats to get the disk usage (although not very useful as we have the metrics independently, might be reused later...).

We also update the LogConsumer class, that was not migrated.

Issue: ARSN-454

Seragonia committed:
Update mongodb and mongodb for tests

Issue: ARSN-454

Seragonia committed:
Update LogConsumer usage of MongoDB driver

- Removed `useNewUrlParser` option from MongoClient instantiation.
- Updated cursor handling to use promises for better error management in fetching the latest oplog ID.
- Ensured proper error handling when no oplog entries are found.

Issue: ARSN-454

Seragonia committed:
Update MongoDB storage engine from 'ephemeralForTest' to 'wiredTiger'

- Modified multiple test files to use 'wiredTiger' as the storage engine for MongoMemoryReplSet.
- This change ensures consistency in the testing environment and aligns with the current best practices for MongoDB testing.
- Previous option was removed in the latest version.

Issue: ARSN-454

Seragonia committed:
Remove useless bson dependency

- We can use mongodb directly

Issue: ARSN-454

Seragonia committed:
Refactor getDiskUsage method in MongoClientInterface

- Updated the getDiskUsage method to retrieve disk usage information directly from MongoDB using the dbStats command.
- Improved error handling for database connection issues and logging for better debugging.
- Removed the dependency on the diskusage package.

Issue: ARSN-454

Seragonia committed:
Refactor MongoClientInterface imports for clarity

- Consolidated MongoDB imports by separating type imports from value imports for better readability and organization.

Issue: ARSN-454

Seragonia committed:
Add getDiskUsage utility and refactor disk usage retrieval

- The diskusage is only relevant for non-unix systems (that we do not support, and won't). We can get rid of this native dependency.
- Introduced a new utility function `getDiskUsage` in `utils.js` to fetch disk usage information using Node.js's native `fs.statfs`.
- Updated `getDiskUsage` methods in `DataFileStore` and `MetadataFileServer` to utilize the new utility function, replacing the previous dependency on the `diskusage` package.
- Improved code organization and clarity by consolidating disk usage logic.

Issue: ARSN-454

Seragonia committed:
Move eslint/plugin-kit from dep to devdep

- Not used during runtime.

Issue: ARSN-454

Seragonia committed:
Move utf8 and async types from dep to dev deps

Issue: ARSN-454

Seragonia committed:
Add unit tests for getDiskUsage utility

- Introduced tests for the `getDiskUsage` function to validate error handling when `fs.statfs` is unavailable or fails.
- Added tests to ensure correct disk usage calculations based on different block sizes.

Issue: ARSN-454

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/arsenal/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue ARSN-454 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.2.14`


Please check the `Fix Version/s` of ARSN-454, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

Seragonia committed:
Close mongodb connection in LogConsumer and add tests

- Added a `close` method to the `LogConsumer` class to properly close the MongoDB connection and handle errors during closure.
- Introduced unit tests for the `LogConsumer` to validate connection handling, including successful closure and error scenarios.
- Created a new test file for `LogConsumer` to ensure comprehensive coverage of its functionality.

Issue: ARSN-454

francoisferrand commented on lib/storage/metadata/mongoclient/LogConsumer.js:
```diff
@@ -89,18 +91,17 @@ class LogConsumer {
-                    cursor, this._logger, startSeq.uniqID, latestOplogID);
-                return cb(null, { log: recordStream, tailable: true });
             });
+            const recordStream = new ListRecordStream(
```
should `ListRecordStream` not be updated as well, to use the new custer?
looking at https://github.dev/scality/Arsenal/blob/dd9283d156a5d7fa57987e48314aa3f904f4508a/lib/storage/metadata/mongoclient/ListRecordStream.js#L47, we see it is still using the callback-based API...

williamlardier replied:
Yes, I can include it in this PR as well...  I add some commits.

williamlardier replied:
Done in https://github.com/scality/Arsenal/pull/2351/commits/44f4beaa759ef9d62968c9e137a23e74927535e3

williamlardier replied:
This whole logic will be removed in ARSN-486.

Seragonia committed:
Refactor ListRecordStream to use async/await

- Updated the `_read` method in `ListRecordStream` to utilize async/await for improved error handling and readability.
- Modified the `next` method in the `MongoCursorMock` class to support async behavior, throwing errors directly instead of using callbacks.

Issue: ARSN-454

ghost commented:
/approve

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.2`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.1`



Please check the status of the associated issue ARSN-454.

Goodbye williamlardier.




*The following options are set:* **approve**
